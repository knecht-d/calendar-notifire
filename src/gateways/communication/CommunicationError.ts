export class CommunicationError extends Error {
    constructor(
        public key: CommunicationErrorCode,
        public given: string = "",
        public expected: string = "",
        public example?: string,
    ) {
        super(`${key}
        Given: ${given}
        Expected: ${expected}
        ${example ? `Example: ${example}` : ""}
        `);

        // Set the prototype explicitly.
        Object.setPrototypeOf(this, CommunicationError.prototype);
    }
}

export enum CommunicationErrorCode {
    MISSING_TRIGGER_ID = "MISSING_TRIGGER_ID",
    INVALID_RECURRENCE_TYPE = "INVALID_RECURRENCE_TYPE",
    INVALID_NUMBER_OF_ARGUMENTS = "INVALID_NUMBER_OF_ARGUMENTS",
    INVALID_DAYS = "INVALID_DAYS",
    INVALID_DAY_OF_MONTH = "INVALID_DAY_OF_MONTH",
    INVALID_TIME = "INVALID_TIME",
    INVALID_FRAME_CONFIG = "INVALID_FRAME_CONFIG",
    INVALID_CRON_SETTING = "INVALID_CRON_SETTING",
}
