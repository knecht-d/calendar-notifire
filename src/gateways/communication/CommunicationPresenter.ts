import { ICommunication, IMessage, MessageKey } from "../../useCases";
import { GateWay } from "../GateWay";
import { CommunicationError } from "./CommunicationError";
import { Mappings } from "./Mappings";

export interface ICommunicationOut {
    send: (chatId: string, message: string) => void;
}

export interface IErrorReporter {
    sendError: (chatId: string, message: string) => void;
    sendCommunicationError: (chatId: string, error: CommunicationError) => void;
}

interface IDependencies {
    communication: ICommunicationOut;
}

export class CommunicationPresenter extends GateWay<IDependencies> implements IErrorReporter, ICommunication {
    sendCommunicationError(chatId: string, error: CommunicationError) {
        const message = this.replacePlaceHolders(Mappings.errorCodes[error.key], {
            given: error.given,
            expected: error.expected,
            example: error.example || "",
        });
        this.sendError(chatId, message);
    }

    sendError(chatId: string, message: string) {
        this.dependencies!.communication.send(chatId, `Fehler: ${message}`);
    }

    send(chatId: string, message: IMessage) {
        this.checkInitialized();
        const text = this.getText(message);
        this.dependencies!.communication.send(chatId, text);
    }

    private getText(message: IMessage): string {
        const mapping = message.hasError ? Mappings.errorMessages : Mappings.successMessages;
        let replacements: { [key: string]: string } = {};
        switch (message.key) {
            case MessageKey.SET_CONFIG:
            case MessageKey.DELETE_CONFIG:
                replacements = { triggerId: message.triggerId };
                break;
            case MessageKey.READ_CONFIG:
                replacements = { timeFrames: JSON.stringify(message.timeFrames, null, "  ") };
                break;
            case MessageKey.INITIALIZE_CHAT:
                replacements = {};
                break;
            case MessageKey.EVENTS:
                replacements = { events: JSON.stringify(message.events, null, "  ") };
                break;
        }
        replacements = {
            message: message.message ? ` ${message.message}` : "",
            ...replacements,
        };
        return this.replacePlaceHolders(mapping[message.key], replacements);
    }

    private replacePlaceHolders(message: string, replacements: { [key: string]: string }) {
        let newMessage = message;
        Object.entries(replacements).forEach(([key, value]) => {
            newMessage = newMessage.replace(new RegExp(`{${key}}`, "gm"), value);
        });
        return newMessage;
    }
}
