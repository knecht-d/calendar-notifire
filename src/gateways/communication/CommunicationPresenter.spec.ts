import { CommunicationPresenter } from "./CommunicationPresenter";
import { CommunicationError } from "./CommunicationError";
// import { Mappings } from "./Mappings";

jest.mock("./Mappings", () => {
    // Works and lets you check for constructor calls:
    return {
        Mappings: {
            errorCodes: {
                GIVEN_EXPECTED_EXAMPLE: "Some Code: Given [{given}] Expected: {expected} Example: {example}",
                NO_EXAMPLE: "Some Code: Given [{given}] Expected: {expected}",
                NO_EXPECTED: "Some Code: Given [{given}] Example: {example}",
                DUPLICATE_GIVEN: "Some Code: Given [{given}] Expected: {expected} Given [{given}]",
            },
        },
    };
});
describe("CommunicationPresenter", () => {
    const mockCommunicationOut = {
        send: jest.fn(),
    };
    beforeEach(() => {
        mockCommunicationOut.send.mockClear();
    });
    describe("sendUpdateSuccess", () => {
        it("should send a plain success message to the chat", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendUpdateSuccess("someChat", "someTrigger");
            expect(mockCommunicationOut.send).toHaveBeenCalledWith("someChat", "Update someTrigger erfolgreich.");
        });
        it("should add the additional message", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendUpdateSuccess("someChat", "someTrigger", "Details");
            expect(mockCommunicationOut.send).toHaveBeenCalledWith(
                "someChat",
                "Update someTrigger erfolgreich. Details",
            );
        });
    });
    describe("sendUpdateError", () => {
        it("should send a plain error message to the chat", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendUpdateError("someChat", "someTrigger");
            expect(mockCommunicationOut.send).toHaveBeenCalledWith(
                "someChat",
                "Fehler: Update someTrigger fehlgeschlagen.",
            );
        });
        it("should add the additional message", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendUpdateError("someChat", "someTrigger", "Details");
            expect(mockCommunicationOut.send).toHaveBeenCalledWith(
                "someChat",
                "Fehler: Update someTrigger fehlgeschlagen - Details",
            );
        });
    });
    describe("sendError", () => {
        it("should send error with 'Fehler:'", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendError("someChat", "reason");
            expect(mockCommunicationOut.send).toHaveBeenCalledWith("someChat", "Fehler: reason");
        });
    });
    describe("sendCommunicationError", () => {
        it("should replace all properties", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendCommunicationError(
                "someChat",
                new CommunicationError("GIVEN_EXPECTED_EXAMPLE" as any, "-given-", "-expected-", "-example-"),
            );
            expect(mockCommunicationOut.send).toHaveBeenCalledWith(
                "someChat",
                "Fehler: Some Code: Given [-given-] Expected: -expected- Example: -example-",
            );
        });
        it("should replace only given and expected", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendCommunicationError(
                "someChat",
                new CommunicationError("NO_EXAMPLE" as any, "-given-", "-expected-"),
            );
            expect(mockCommunicationOut.send).toHaveBeenCalledWith(
                "someChat",
                "Fehler: Some Code: Given [-given-] Expected: -expected-",
            );
        });
        it("should only replace the placeholders", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendCommunicationError(
                "someChat",
                new CommunicationError("NO_EXPECTED" as any, "-given-", "-expected-", "-example-"),
            );
            expect(mockCommunicationOut.send).toHaveBeenCalledWith(
                "someChat",
                "Fehler: Some Code: Given [-given-] Example: -example-",
            );
        });
        it("should only replace all duplicated placeholders", () => {
            const presenter = new CommunicationPresenter(mockCommunicationOut);
            presenter.sendCommunicationError(
                "someChat",
                new CommunicationError("DUPLICATE_GIVEN" as any, "-given-", "-expected-", "-example-"),
            );
            expect(mockCommunicationOut.send).toHaveBeenCalledWith(
                "someChat",
                "Fehler: Some Code: Given [-given-] Expected: -expected- Given [-given-]",
            );
        });
    });
});
