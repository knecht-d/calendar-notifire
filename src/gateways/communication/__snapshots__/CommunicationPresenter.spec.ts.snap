// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommunicationPresenter sendEvents should send the events 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "someChat",
      "Events: [
  {
    \\"start\\": \\"2020-05-01T10:00:00.000Z\\",
    \\"end\\": \\"2020-05-01T10:00:00.000Z\\",
    \\"title\\": \\"Event\\"
  }
]",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
